tracker/clientcontext/injector.go

```diff
@@ -0,0 +1,179 @@
+	ctx context.Context,
+	conn N.PacketConn,
+	metadata adapter.InboundContext,
+	matchedRule adapter.Rule,
+	matchOutbound adapter.Outbound,
+) N.PacketConn {
+	if !t.match(metadata.Inbound, matchOutbound.Tag()) {
+		return conn
+	}
+	info := t.getInfo()
+	return newWritePacketConn(conn, metadata, &info)
+}
+
+func (t *ClientContextInjector) match(inbound, outbound string) bool {
+	t.ruleMu.RLock()
+	defer t.ruleMu.RUnlock()
+	return t.inboundRule.match(inbound) && t.outboundRule.match(outbound)
+}
+
+func (t *ClientContextInjector) UpdateBounds(bounds MatchBounds) {
+	t.ruleMu.Lock()
+	t.inboundRule = newBoundsRule(bounds.Inbound)
+	t.outboundRule = newBoundsRule(bounds.Outbound)
+	t.ruleMu.Unlock()
+}
+
+// writeConn sends client info after handshake.
+type writeConn struct {
+	net.Conn
+	info *ClientInfo
+}
+
+func newWriteConn(conn net.Conn, info *ClientInfo) net.Conn {
+	return &writeConn{Conn: conn, info: info}
+}
+
+// ConnHandshakeSuccess sends client info upon successful handshake with the server.
+func (c *writeConn) ConnHandshakeSuccess(conn net.Conn) error {
+	if err := c.sendInfo(conn); err != nil {
+		return fmt.Errorf("sending client info: %w", err)
+	}
+	return nil
+}
+
+// sendInfo marshals and sends client info as an HTTP POST, then waits for HTTP 200 OK.
+func (c *writeConn) sendInfo(conn net.Conn) error {
+	buf, err := json.Marshal(c.info)
+	if err != nil {
+		return fmt.Errorf("marshaling client info: %w", err)
+	}
+	// Write HTTP POST request
+	req := bytes.NewBuffer(nil)
+	fmt.Fprintf(req, "POST /clientinfo HTTP/1.1\r\n")
+	fmt.Fprintf(req, "Host: localhost\r\n")
+	fmt.Fprintf(req, "Content-Type: application/json\r\n")
+	fmt.Fprintf(req, "Content-Length: %d\r\n", len(buf))
+	fmt.Fprintf(req, "\r\n")
+	req.Write(buf)
+	if _, err = conn.Write(req.Bytes()); err != nil {
+		return fmt.Errorf("writing client info: %w", err)
+	}
+
+	// wait for HTTP 200 OK response
+	reader := stdbufio.NewReader(conn)
+	resp, err := http.ReadResponse(reader, nil)
+	if err != nil {
+		return fmt.Errorf("reading HTTP response: %w", err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("invalid server response: %s", resp.Status)
+	}
+	return nil
+}
+
+type writePacketConn struct {
+	N.PacketConn
+	metadata adapter.InboundContext
+	info     *ClientInfo
+}
+
+func newWritePacketConn(
+	conn N.PacketConn,
+	metadata adapter.InboundContext,
+	info *ClientInfo,
+) N.PacketConn {
+	return &writePacketConn{
+		PacketConn: conn,
+		metadata:   metadata,
+		info:       info,
+	}
+}
+
+// PacketConnHandshakeSuccess sends client info upon successful handshake.
+func (c *writePacketConn) PacketConnHandshakeSuccess(conn net.PacketConn) error {
+	if err := c.sendInfo(conn); err != nil {
+		return fmt.Errorf("sending client info: %w", err)
+	}
+	return nil
+}
+
+// sendInfo marshals and sends client info as a CLIENTINFO packet, then waits for OK.
+func (c *writePacketConn) sendInfo(conn net.PacketConn) error {
```

Comment:
Why are the sendInfo methods different for `net.Conn` and `net.PacketConn`? Are you worried about going over the MTU?